File: flutter/storyly_flutter/android/src/main/kotlin/com/appsamurai/storyly/storyly_flutter/vertical_feed/FlutterVerticalFeedPresenterView.kt
package com.appsamurai.storyly.storyly_flutter.vertical_feed


import android.content.Context
import android.view.View
import com.appsamurai.storyly.StorylyDataSource
import com.appsamurai.storyly.StorylyListener
import com.appsamurai.storyly.analytics.VerticalFeedEvent
import com.appsamurai.storyly.data.managers.product.STRCart
import com.appsamurai.storyly.data.managers.product.STRCartEventResult
import com.appsamurai.storyly.data.managers.product.STRCartItem
import com.appsamurai.storyly.data.managers.product.STRProductInformation
import com.appsamurai.storyly.storyly_flutter.createSTRCart
import com.appsamurai.storyly.storyly_flutter.createSTRCartItemMap
import com.appsamurai.storyly.storyly_flutter.createSTRCartMap
import com.appsamurai.storyly.storyly_flutter.createSTRProductInformationMap
import com.appsamurai.storyly.storyly_flutter.createSTRProductItem
import com.appsamurai.storyly.verticalfeed.StorylyVerticalFeedPresenterView
import com.appsamurai.storyly.verticalfeed.VerticalFeedGroup
import com.appsamurai.storyly.verticalfeed.VerticalFeedItem
import com.appsamurai.storyly.verticalfeed.VerticalFeedItemComponent
import com.appsamurai.storyly.verticalfeed.listener.StorylyVerticalFeedPresenterListener
import com.appsamurai.storyly.verticalfeed.listener.StorylyVerticalFeedPresenterProductListener
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import java.util.UUID


class FlutterVerticalFeedPresenterViewFactory(private val messenger: BinaryMessenger) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    internal lateinit var context: Context

    override fun create(context: Context?, viewId: Int, args: Any?): PlatformView = FlutterVerticalFeedPresenterView(this.context, messenger, viewId, args as HashMap<String, Any>)
}

class FlutterVerticalFeedPresenterView(
    private val context: Context,
    messenger: BinaryMessenger,
    viewId: Int,
    private val args: HashMap<String, Any>
) : PlatformView, StorylyListener {

    private var cartUpdateSuccessFailCallbackMap: MutableMap<String, Pair<((STRCart?) -> Unit)?, ((STRCartEventResult) -> Unit)?>> = mutableMapOf()

    private val methodChannel: MethodChannel = MethodChannel(messenger, "com.appsamurai.storyly/flutter_vertical_feed_presenter_$viewId").apply {
        setMethodCallHandler { call, _ ->
            val callArguments = call.arguments as? Map<String, *>
            when (call.method) {
                "refresh" -> verticalFeedView.refresh()
                "pause" -> verticalFeedView.pause()
                "play" -> verticalFeedView.play()
                "hydrateProducts" -> (callArguments?.get("products") as? List<Map<String, Any?>>)?.let {
                    val products = it.map { product -> createSTRProductItem(product) }
                    verticalFeedView.hydrateProducts(products)
                }

                "updateCart" -> (callArguments?.get("cart") as? Map<String, Any?>)?.let {
                    verticalFeedView.updateCart(createSTRCart(it))
                }

                "approveCartChange" -> (callArguments?.get("responseId") as? String)?.let {
                    val onSuccess = cartUpdateSuccessFailCallbackMap[it]?.first
                    (callArguments["cart"] as? Map<String, Any?>)?.let { cartMap ->
                        onSuccess?.invoke(createSTRCart(cartMap))
                    } ?: kotlin.run {
                        onSuccess?.invoke(null)
                    }
                    cartUpdateSuccessFailCallbackMap.remove(it)
                }

                "rejectCartChange" -> (callArguments?.get("responseId") as? String)?.let {
                    val onFail = cartUpdateSuccessFailCallbackMap[it]?.second
                    (callArguments["failMessage"] as? String)?.let { failMessage ->
                        onFail?.invoke(STRCartEventResult(failMessage))
                    }
                    cartUpdateSuccessFailCallbackMap.remove(it)
                }
            }
        }
    }

    private val verticalFeedView: StorylyVerticalFeedPresenterView by lazy {
        StorylyVerticalFeedPresenterView(context).apply {
            storylyVerticalFeedInit = VerticalFeedInitMapper(context).getStorylyInit(json = args) ?: return@apply

            storylyVerticalFeedProductListener = object : StorylyVerticalFeedPresenterProductListener {
                override fun verticalFeedEvent(
                    view: StorylyVerticalFeedPresenterView,
                    event: VerticalFeedEvent
                ) {
                    methodChannel.invokeMethod(
                        "verticalFeedEvent",
                        mapOf(
                            "event" to event.name,
                        )
                    )
                }

                override fun verticalFeedHydration(
                    view: StorylyVerticalFeedPresenterView,
                    products: List<STRProductInformation>
                ) {
                    methodChannel.invokeMethod(
                        "verticalFeedOnHydration",
                        mapOf(
                            "products" to products.map { createSTRProductInformationMap(it) }
                        )
                    )
                }

                override fun verticalFeedUpdateCartEvent(
                    view: StorylyVerticalFeedPresenterView,
                    event: VerticalFeedEvent,
                    cart: STRCart?,
                    change: STRCartItem?,
                    onSuccess: ((STRCart?) -> Unit)?,
                    onFail: ((STRCartEventResult) -> Unit)?
                ) {
                    val responseId = UUID.randomUUID().toString()
                    cartUpdateSuccessFailCallbackMap[responseId] = Pair(onSuccess, onFail)

                    methodChannel.invokeMethod(
                        "verticalFeedOnProductCartUpdated",
                        mapOf(
                            "event" to event.name,
                            "cart" to createSTRCartMap(cart),
                            "change" to createSTRCartItemMap(change),
                            "responseId" to responseId
                        )
                    )
                }
            }

            storylyVerticalFeedListener = object : StorylyVerticalFeedPresenterListener {

                override fun verticalFeedLoaded(
                    view: StorylyVerticalFeedPresenterView,
                    feedGroupList: List<VerticalFeedGroup>,
                    dataSource: StorylyDataSource
                ) {
                    methodChannel.invokeMethod(
                        "verticalFeedLoaded",
                        mapOf(
                            "feedGroupList" to feedGroupList.map { storyGroup -> createStoryGroupMap(storyGroup) },
                            "dataSource" to dataSource.value
                        )
                    )
                }

                override fun verticalFeedLoadFailed(
                    view: StorylyVerticalFeedPresenterView,
                    errorMessage: String
                ) {
                    methodChannel.invokeMethod("verticalFeedLoadFailed", errorMessage)
                }

                override fun verticalFeedEvent(
                    view: StorylyVerticalFeedPresenterView,
                    event: VerticalFeedEvent,
                    feedGroup: VerticalFeedGroup?,
                    feedItem: VerticalFeedItem?,
                    feedItemComponent: VerticalFeedItemComponent?
                ) {
                    methodChannel.invokeMethod(
                        "verticalFeedEvent",
                        mapOf("event" to event.name,
                            "feedGroup" to feedGroup?.let { createStoryGroupMap(feedGroup) },
                            "feedItem" to feedItem?.let { createStoryMap(feedItem) },
                            "feedItemComponent" to feedItemComponent?.let { createStoryComponentMap(feedItemComponent) })
                    )
                }

                override fun verticalFeedShown(view: StorylyVerticalFeedPresenterView) {
                    methodChannel.invokeMethod("verticalFeedShown", null)
                }

                override fun verticalFeedShowFailed(
                    view: StorylyVerticalFeedPresenterView,
                    errorMessage: String
                ) {
                    methodChannel.invokeMethod("verticalFeedShowFailed", errorMessage)
                }

                override fun verticalFeedDismissed(view: StorylyVerticalFeedPresenterView) {
                    methodChannel.invokeMethod("verticalFeedDismissed", null)
                }

                override fun verticalFeedActionClicked(
                    view: StorylyVerticalFeedPresenterView,
                    feedItem: VerticalFeedItem
                ) {
                    methodChannel.invokeMethod(
                        "verticalFeedActionClicked",
                        createStoryMap(feedItem)
                    )
                }

                override fun verticalFeedUserInteracted(
                    view: StorylyVerticalFeedPresenterView,
                    feedGroup: VerticalFeedGroup,
                    feedItem: VerticalFeedItem,
                    feedItemComponent: VerticalFeedItemComponent
                ) {
                    methodChannel.invokeMethod(
                        "verticalFeedUserInteracted",
                        mapOf(
                            "feedGroup" to createStoryGroupMap(feedGroup),
                            "feedItem" to createStoryMap(feedItem),
                            "feedItemComponent" to createStoryComponentMap(feedItemComponent)
                        )
                    )
                }
            }
        }
    }

    override fun getView(): View = verticalFeedView

    override fun dispose() {}
}

File: legacy/flutter/storyly_flutter/ios/Classes/FlutterVerticalFeedPresenterView.swift
import UIKit
import Storyly


public class FlutterVerticalFeedPresenterViewFactory: NSObject, FlutterPlatformViewFactory {
    private let messenger: FlutterBinaryMessenger
    private var viewList: [Int64: Weak<FlutterVerticalFeedPresenterView>] = [:]
    
    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
    }
    
    public func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        let flutterStorylyView =  FlutterVerticalFeedPresenterView(frame, viewId: viewId, args: args as? [String : Any] ?? [:], messenger: self.messenger)
        self.viewList[viewId] = Weak<FlutterVerticalFeedPresenterView>(value: flutterStorylyView)
        flutterStorylyView.onDispose = { [weak self] viewId in
            self?.viewList.removeValue(forKey: viewId)
        }
        return flutterStorylyView
    }
    
    public func getViewById(id: Int) -> StorylyVerticalFeedPresenterView? {
        guard let view = viewList[Int64(id)] else { return nil }
        return view.value?.getStorylyView()
    }
    
    public func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

public class FlutterVerticalFeedPresenterView: NSObject, FlutterPlatformView {
    private let frame: CGRect
    private let viewId: Int64
    private let args: [String: Any]
    private let methodChannel: FlutterMethodChannel
    
    internal var onDispose: ((_ viewId: Int64) -> ())? = nil
    
    private lazy var storylyViewWrapper = FlutterVerticalFeedViewPresenterWrapper(frame: self.frame, args: self.args, methodChannel: self.methodChannel)
    
    init(_ frame: CGRect, viewId: Int64, args: [String: Any], messenger: FlutterBinaryMessenger) {
        self.frame = frame
        self.viewId = viewId
        self.args = args
        self.methodChannel = FlutterMethodChannel(name: "com.appsamurai.storyly/flutter_vertical_feed_presenter_\(viewId)", binaryMessenger: messenger)
    }
    
    public func getStorylyView() -> StorylyVerticalFeedPresenterView {
        return storylyViewWrapper.verticalFeedView
    }
    
    public func view() -> UIView {
        return self.storylyViewWrapper
    }
    
    deinit {
        onDispose?(viewId)
    }
}

internal class FlutterVerticalFeedViewPresenterWrapper: UIView {
    internal let verticalFeedView: StorylyVerticalFeedPresenterView
    
    private let methodChannel: FlutterMethodChannel
    
    private var cartUpdateSuccessFailCallbackMap: [String: (((STRCart?) -> Void)?, ((STRCartEventResult) -> Void)?)] = [:]
    private var wishlistUpdateSuccessFailCallbackMap: [String: (((STRProductItem?) -> Void)?, ((STRWishlistEventResult) -> Void)?)] = [:]

    init(frame: CGRect,
         args: [String: Any],
         methodChannel: FlutterMethodChannel) {
        self.methodChannel = methodChannel
        self.verticalFeedView = StorylyVerticalFeedPresenterView(frame: frame)
        super.init(frame: frame)
        
        self.methodChannel.setMethodCallHandler { [weak self] call, _ in
            guard let self = self else { return }
            let callArguments = call.arguments as? [String: Any]
            switch call.method {
            case "refresh": self.verticalFeedView.refresh()
            case "play": self.verticalFeedView.play()
            case "pause": self.verticalFeedView.pause()
            case "hydrateProducts":
                if let products = callArguments?["products"] as? [[String : Any?]] {
                    let storylyProducts = products.compactMap { createSTRProductItem(product: $0) }
                    self.verticalFeedView.hydrateProducts(products: storylyProducts)
                }
            case "hydrateWishlist":
                if let products = callArguments?["products"] as? [[String : Any?]] {
                    let storylyProducts = products.compactMap { createSTRProductItem(product: $0) }
                    self.verticalFeedView.hydrateWishlist(products: storylyProducts)
                }    
            case "updateCart":
                if let cart = callArguments?["cart"] as? [String : Any?] {
                    self.verticalFeedView.updateCart(cart: createSTRCart(cartMap: cart))
                }
            case "approveCartChange":
                if let responseId = callArguments?["responseId"] as? String,
                   let onSuccess = self.cartUpdateSuccessFailCallbackMap[responseId]?.0 as? (STRCart?) -> Void {
                    if let cart = callArguments?["cart"] as? [String : Any?] {
                        onSuccess(createSTRCart(cartMap: cart))
                    } else {
                        onSuccess(nil)
                    }
                    self.cartUpdateSuccessFailCallbackMap.removeValue(forKey: responseId)
                }
            case "rejectCartChange":
                if let responseId = callArguments?["responseId"] as? String,
                   let onFail = self.cartUpdateSuccessFailCallbackMap[responseId]?.1 as? (STRCartEventResult) -> Void,
                   let failMessage = callArguments?["failMessage"] as? String {
                    onFail(STRCartEventResult(message: failMessage))
                    self.cartUpdateSuccessFailCallbackMap.removeValue(forKey: responseId)
                }
            case "approveWishlistChange":
                if let responseId = callArguments?["responseId"] as? String,
                   let onSuccess = self.wishlistUpdateSuccessFailCallbackMap[responseId]?.0 as? (STRProductItem?) -> Void {
                    if let item = callArguments?["item"] as? [String : Any?] {
                        onSuccess(createSTRProductItem(product: item))
                    } else {
                        onSuccess(nil)
                    }
                    self.wishlistUpdateSuccessFailCallbackMap.removeValue(forKey: responseId)
                }
            case "rejectWishlistChange":
                if let responseId = callArguments?["responseId"] as? String,
                   let onFail = self.wishlistUpdateSuccessFailCallbackMap[responseId]?.1 as? (STRWishlistEventResult) -> Void,
                   let failMessage = callArguments?["failMessage"] as? String {
                    onFail(STRWishlistEventResult(message: failMessage))
                    self.wishlistUpdateSuccessFailCallbackMap.removeValue(forKey: responseId)
                }
            default: do {}
            }
        }
        
        guard let storylyInit = VerticalFeedInitMapper().getStorylyInit(json: args) else { return }
        self.verticalFeedView.translatesAutoresizingMaskIntoConstraints = false
        self.verticalFeedView.storylyVerticalFeedInit = storylyInit
        self.verticalFeedView.storylyVerticalFeedDelegate = self
        self.verticalFeedView.storylyVerticalFeedProductDelegate = self
        self.verticalFeedView.rootViewController = UIApplication.shared.keyWindow?.rootViewController?.getPresentedViewController()
        self.addSubview(verticalFeedView)
        
        self.verticalFeedView.leadingAnchor.constraint(equalTo: self.leadingAnchor).isActive = true
        self.verticalFeedView.trailingAnchor.constraint(equalTo: self.trailingAnchor).isActive = true
        self.verticalFeedView.topAnchor.constraint(equalTo: self.topAnchor).isActive = true
        self.verticalFeedView.bottomAnchor.constraint(equalTo: self.bottomAnchor).isActive = true
    }
    
    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }
    
}

extension FlutterVerticalFeedViewPresenterWrapper: StorylyVerticalFeedPresenterProductDelegate {
    
    func verticalFeedEvent(
        _ view: StorylyVerticalFeedPresenterView,
        event: VerticalFeedEvent
    ) {
        
        methodChannel.invokeMethod(
            "verticalFeedProductEvent",
            arguments: ["event": event.stringValue]
        )
    }
    
    func verticalFeedHydration(
        _ view: StorylyVerticalFeedPresenterView,
        products: [STRProductInformation]
    ) {
        methodChannel.invokeMethod(
            "verticalFeedOnProductHydration",
            arguments: [
                "products": products.map { createSTRProductInformationMap(productInfo: $0) }
            ]
        )
    }
    
    func verticalFeedUpdateCartEvent(
        view: StorylyVerticalFeedPresenterView,
        event: VerticalFeedEvent,
        cart: STRCart?,
        change: STRCartItem?,
        onSuccess: ((STRCart?) -> Void)?,
        onFail: ((STRCartEventResult) -> Void)?
    ) {
        let responseId = UUID().uuidString
        cartUpdateSuccessFailCallbackMap[responseId] = (onSuccess, onFail)
        
        methodChannel.invokeMethod(
            "verticalFeedOnProductCartUpdated",
            arguments: [
                "event": event.stringValue,
                "cart": createSTRCartMap(cart: cart),
                "change": createSTRCartItemMap(cartItem: change),
                "responseId": responseId
            ]  as [String: Any?]
        )
    }
    
    func verticalFeedUpdateWishlistEvent(view: StorylyVerticalFeedPresenterView, item: STRProductItem?, event: StorylyEvent, onSuccess: ((STRProductItem?) -> Void)?, onFail: ((STRWishlistEventResult) -> Void)?) {
            let responseId = UUID().uuidString
            self.wishlistUpdateSuccessFailCallbackMap[responseId] = (onSuccess, onFail)
            
            self.methodChannel.invokeMethod(
                "verticalFeedOnWishlistUpdated",
                arguments: [
                    "event": event.stringValue,
                    "item": createSTRProductItemMap(product: item),
                    "responseId": responseId
                ]
            )
    }
}


extension FlutterVerticalFeedViewPresenterWrapper: StorylyVerticalFeedPresenterDelegate {
    func verticalFeedLoaded(
        _ view: StorylyVerticalFeedPresenterView,
        feedGroupList: [VerticalFeedGroup],
        dataSource: StorylyDataSource
    ) {
        methodChannel.invokeMethod(
            "verticalFeedLoaded",
            arguments: [
                "feedGroupList": feedGroupList.map { createStoryGroupMap(storyGroup: $0) },
                "dataSource": dataSource.description
            ]
        )
    }
    
    func verticalFeedLoadFailed(
        _ view: StorylyVerticalFeedPresenterView,
        errorMessage: String
    ) {
        methodChannel.invokeMethod("verticalFeedLoadFailed", arguments: errorMessage)
    }
    
    func verticalFeedEvent(
        _ view: StorylyVerticalFeedPresenterView,
        event: VerticalFeedEvent,
        feedGroup: VerticalFeedGroup?,
        feedItem: VerticalFeedItem?,
        feedItemComponent: VerticalFeedItemComponent?
    ) {
        methodChannel.invokeMethod(
            "verticalFeedEvent",
            arguments: [
                "event": event.stringValue,
                "feedGroup": createStoryGroupMap(storyGroup: feedGroup),
                "feedItem": createStoryMap(story: feedItem),
                "feedItemComponent": createStoryComponentMap(storyComponent: feedItemComponent)
            ] as [String: Any?]
        )
    }
    
    func verticalFeedPresented(_ view: StorylyVerticalFeedPresenterView) {
        methodChannel.invokeMethod("verticalFeedShown", arguments: nil)
    }
    
    func verticalFeedPresentFailed(
        _ view: StorylyVerticalFeedPresenterView,
        errorMessage: String
    ) {
        methodChannel.invokeMethod("verticalFeedShowFailed", arguments: errorMessage)
    }
    
    func verticalFeedDismissed(_ view: StorylyVerticalFeedPresenterView) {
        methodChannel.invokeMethod("verticalFeedDismissed", arguments: nil)
    }
    
    func verticalFeedActionClicked(
        _ view: StorylyVerticalFeedPresenterView,
        rootViewController: UIViewController,
        feedItem: VerticalFeedItem) {
        methodChannel.invokeMethod(
            "verticalFeedActionClicked",
            arguments: createStoryMap(story: feedItem)
        )
    }
    
    func verticalFeedUserInteracted(
        _ view: StorylyVerticalFeedPresenterView,
        feedGroup: VerticalFeedGroup,
        feedItem: VerticalFeedItem,
        feedItemComponent: VerticalFeedItemComponent
    ) {
        methodChannel.invokeMethod(
            "verticalFeedUserInteracted",
            arguments: [
                "feedGroup": createStoryGroupMap(storyGroup: feedGroup),
                "feedItem": createStoryMap(story: feedItem),
                "feedItemComponent": createStoryComponentMap(storyComponent: feedItemComponent)
            ] as [String: Any?]
        )
    }
}
